build.rs

```diff
@@ -154,6 +154,10 @@ fn write_script<W: io::Write>(w: &mut W) -> Result<()> {
 #
 
 set -e
+
+git stash --keep-index --include-untracked --quiet
+trap "git reset --quiet --hard HEAD && git stash pop --index --quiet" EXIT INT TERM
```

Comment:
I think this `git stash pop` will fail when the script generates some untracked file. For example, let's say `cargo test` generate `output.txt`. If `output.txt` already exists, `git stash push` saves the `output.txt`, then `cargo test` will be run for check which generates new `output.txt`. Finally `git stash pop` tries to restore `output.txt` but it already exists. I think we should clean untracked files before `git stash pop`.

Reply:
I'll respond to the second point above here, since it's the same answer.

The `reset --hard` is needed to avoid `pop` conflicts. Since the hard reset would literally wipe the repo we need `--include-untracked` to save all the files. We would lose all files generated by the hook, true, but I'm not sure this is avoidable

Reply:
> Since the hard reset would literally wipe the repo we need

I think `git reset --hard` does not wipe untracked files. Please try the following in some Git repository

```sh
# Generate a file
echo hello > hello.txt

# Check hello.txt is untracked
git status

# Run hard reset
git reset --hard HEAD

# hello.txt would still exist
```

Reply:
Damn, my git-fu is failing me! Removed the reset and the `--include-untracked`. Not sure how to handle the cargo commands generating files.